modules/proposals/engine/src/lib.rs

```diff
@@ -274,91 +316,233 @@ impl<T: Trait> Module<T> {
         <system::Module<T>>::block_number()
     }
 
+    // Enumerates through active proposals. Tally Voting results.
+    // Returns proposals with finalized status and id
+    fn get_finalized_proposals() -> Vec<FinalizedProposal<T>> {
```

Comment:
This routine, and the associated types, was really hard to grapple with. Some of it is undoubtedly due to some names that are suboptimal, such as `FinalizedProposal` not actually being a proposal, but an alias for `FinalizedProposalData`, despite the name `get_finalized_proposals`. Also `Proposal::define_proposal_decision_status` does not give a `ProposalDecisionStatus`, but an option of this. On top of this, there are private types like `ProposalStatusResolution`, which needs both a `Proposal`and specific fields in the proposal, like `approvals`, `slashes` etc.

In general, this is all to say, I still can't fully work out if this is entirely correct and efficient, but I suspect it is, do you see any way to simplify and clarify this whole section?